contributors/design-proposals/pod-injection-policy.md

```diff
@@ -0,0 +1,676 @@
+    * [ReplicaSet with Pod Spec Example](#replicaset-with-pod-spec-example)
+    * [Multiple PodInjectionPolicy Example](#multiple-podinjectionpolicy-example)
+    * [Conflict Example](#conflict-example)
+
+
+## Abstract
+
+Describes a policy resource that allows for the loose coupling of a Pod's
+definition from additional runtime requirements for that Pod. For example,
+mounting of Secrets, or setting additional environment variables,
+may not be known at Pod deployment time, but may be required at Pod creation
+time.
+
+## Motivation
+
+Consuming a service involves more than just connectivity. In addition to
+coordinates to reach the service, credentials and non-secret configuration
+parameters are typically needed to use the service.  The primitives for this
+already exist, but a gap exists where loose coupling is desired: it should be
+possible to inject pods with the information they need to use a service on a
+service-by-service basis, without the pod authors having to incorporate the
+information into every pod spec where it is needed.
+
+## Constraints and Assumptions
+
+1.  Future work might require new mechanisms to be made to work with existing
+    controllers such as deployments and replicasets that create pods. Existing
+    controllers that create pods should recreate their pods when a new Pod Injection
+    Policy is added that would effect them.
+
+## Use Cases
+
+1.  As a user, I want to be able to provision an instance of a new service
+    without needing to know the application configuration primitives the
+    service will consume.
+2.  As a cluster admin, I want specific configuration items of a service to be
+    withheld from a developer deploying a service, but not to block the
+    developer from shipping.
+
+### Loose coupling between services and their consumers
+
+Using a Pod Injection Policy allows pod template authors to not have to explicitly
+set information for every pod. This way authors of pod templates consuming a
+specific service do not need to know all the details about that service.
+
+The pods chosen for injection will be based off the policy's `Selector`.
+
+**Example:**
+
+1. Database Administrator provisions a MySQL service for their cluster.
+2. Database Administrator creates secrets for the cluster containing the
+   database name, username, and password.
+3. Database Administrator creates a `PodInjectionPolicy`  defining the database
+   port as an enviornment variable, as well as the secrets. See
+   [Examples](#examples) below for various examples.
+4. Developer of an application can now label their service with the specified
+   `Selector` the Database Administrator tells them, and consume the MySQL
+   database without needing to know any of the details from step 2 and 3.
+
+## Proposed Changes
+
+### PodInjectionPolicy API object
+
+This resource is alpha. The policy itself is immutable. The API group will be
+added to `apps` and the version is `v1alpha1`.
```

Comment:
I don't have a better suggestion right now, but almost certainly I'm going to ask this to be moved to a different group, since I envision it at a different layer of the system than the other APIs in the apps group.

https://docs.google.com/document/d/1XkjVm4bOeiVkj-Xt1LgoGiqWsBfNozJ51dyI-ljzt1o/edit